### PR TITLE
Improve performance of `merge_leaf_stats()`

### PR DESCRIPTION
**Pull request by asubramanya**
The previous algorithm to merge the HLL counters for leaf partitions had a time
complexity of O(n^2) where n is the number of leaf partitions. In case where
the number of partitions was high, resulting in poor performance of analyze on
partitioned tables.  This patch fixes the performance issue by using a new
algorithm to merge the leaf HLL counters in O(n) time.

Co-authored-by: Omer Arap <oarap@pivotal.io>
Co-authored-by: Bhuvnesh Chaudhary <bchaudhary@pivotal.io>

## Here are some reminders before you submit the pull request
- [] Add tests for the change
- [] Document changes
- [] Communicate in the mailing list if needed
- [] Pass `make installcheck`

### CODE REVIEW

**Comment by pivotal-ning-yu** [comment]
On Saturday, December 15, 2018, Abhijit Subramanya <notifications@github.com>
wrote:

> The previous algorithm to merge the HLL counters for leaf partitions had a
> time
> complexity of O(n^2) where n is the number of leaf partitions. In case
> where
> the number of partitions was high, resulting in poor performance of
> analyze on
> partitioned tables. This patch fixes the performance issue by using a new
> algorithm to merge the leaf HLL counters in O(n) time.
>

is there benchmark data on the performance? I am not familar with this part
of logic, but a quick glance on the patch makes me wondering the complexity
of hyperloglog_merge_counters , is it O(1)? I didn't have the source code
at hand at moment, sorry I cannot check it myself.

Co-authored-by: Omer Arap oarap@pivotal.io

File: src/backend/commands/analyze.c [hunk 1]
```diff
@@ -4094,39 +4094,95 @@ merge_leaf_stats(VacAttrStatsP stats,
...
-						continue;
-
-					HLLCounter finalHLL_NDV = NULL;
-					for (j = 0; j < numPartitions; j++)
+					/*
+					 * The first and the last element in the left and right arrays
+					 * are default values since there is no element towards
+					 * the left or right of them
+					 */
+					if (i > 0)
```
Comment (danielgustafsson) [inline src/backend/commands/analyze.c]: If we never consider the `i == 0` case, can't the `for` loop start at one with the comment moved up explaining why? Or am I misunderstanding this logic?

Comment (asubramanya) [reply src/backend/commands/analyze.c]: You are right. We have made the change.

**Comment by asubramanya** [comment]
@pivotal-ning-yu 
Please find below the results of performance testing with and without the fix. We created a partition table with 2 columns, 5000 partitions and loaded each partition with 10,000 rows. This was on my laptop with 3 segment cluster.

Setup:
```
abhijits=# create table part(a int, b int) partition by range(b) (start(1) end(5000) every(1));
abhijits=# insert into part select i, (i % 4999) + 1 from generate_series(1, 50000000)i;
```

With fix:
```
abhijits=# analyze part;
ANALYZE
Time: 62058.287 ms
```

Without fix:
```
abhijits=# analyze part;
... 20 mins gone, Still waiting for it to finish
```

When I mentioned O(n) its for the for loop which goes ahead to merge the counter (not the hyperloglog_merge_counters function). As mentioned in the description, there were 2 nested loops, making it run for n^2 time, but now its runs only n times.

File: src/backend/commands/analyze.c [hunk 2]
```diff
@@ -4094,39 +4094,92 @@ merge_leaf_stats(VacAttrStatsP stats,
...
+
+					pfree(hllcounter_temp1);
+					pfree(hllcounter_temp2);
+
+					if (final != NULL)
 					{
-						// Calculating uniques in each partition
-						nUniques[i] =
-							ndistinct - hyperloglog_estimate(finalHLL_NDV);
+						nUniques[i] = ndistinct - hyperloglog_estimate(final);
```
Comment (chrishajas) [inline src/backend/commands/analyze.c]: Can we just make nUniques a local variable instead of an array? I don't see it referenced anywhere else.

Comment (chrishajas) [reply src/backend/commands/analyze.c]: Looks like the PR pipeline is failing after this change, not sure why though.

Comment (asubramanya) [reply src/backend/commands/analyze.c]: I see a similar failure in the master pipeline too: https://prod.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb_master/jobs/icw_planner_centos6/builds/914
So probably not related.

**Pull request merged**